pkg/webhook/conversion/conversion_test.go

```diff
@@ -0,0 +1,227 @@
+
+	var respRecorder *httptest.ResponseRecorder
+	var decoder *Decoder
+	var scheme *runtime.Scheme
+	webhook := Webhook{}
+
+	BeforeEach(func() {
+		respRecorder = &httptest.ResponseRecorder{
+			Body: bytes.NewBuffer(nil),
+		}
+
+		scheme = kscheme.Scheme
+		Expect(jobsapis.AddToScheme(scheme)).Should(Succeed())
+
+		Expect(webhook.InjectScheme(scheme)).Should(Succeed())
+
+		var err error
+		decoder, err = NewDecoder(scheme)
+		Expect(err).NotTo(HaveOccurred())
+
+	})
+
+	doRequest := func(convReq *apix.ConversionReview) *apix.ConversionReview {
+		var payload bytes.Buffer
+
+		Expect(json.NewEncoder(&payload).Encode(convReq)).Should(Succeed())
+
+		convReview := &apix.ConversionReview{}
+		req := &http.Request{
+			Body: ioutil.NopCloser(bytes.NewReader(payload.Bytes())),
+		}
+		webhook.ServeHTTP(respRecorder, req)
+		Expect(json.NewDecoder(respRecorder.Result().Body).Decode(convReview)).Should(Succeed())
+		return convReview
+	}
+
+	It("should convert objects successfully", func() {
+
+		v1Obj := &jobsv1.ExternalJob{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "ExternalJob",
+				APIVersion: "jobs.example.org/v1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: "default",
+				Name:      "obj-1",
+			},
+			Spec: jobsv1.ExternalJobSpec{
+				RunAt: "every 2 seconds",
+			},
+		}
+
+		expected := &jobsv2.ExternalJob{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "ExternalJob",
+				APIVersion: "jobs.example.org/v2",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: "default",
+				Name:      "obj-1",
+			},
+			Spec: jobsv2.ExternalJobSpec{
+				ScheduleAt: "every 2 seconds",
+			},
+		}
+
+		convReq := &apix.ConversionReview{
+			TypeMeta: metav1.TypeMeta{},
+			Request: &apix.ConversionRequest{
+				DesiredAPIVersion: "jobs.example.org/v2",
+				Objects: []runtime.RawExtension{
+					{
+						Object: v1Obj,
+					},
+				},
+			},
+		}
+
+		convReview := doRequest(convReq)
+
+		Expect(convReview.Response.ConvertedObjects).Should(HaveLen(1))
+		got, _, err := decoder.Decode(convReview.Response.ConvertedObjects[0].Raw)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(got).To(Equal(expected))
+	})
+
+	It("should return error when dest/src objects belong to different API groups", func() {
+		v1Obj := &jobsv1.ExternalJob{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "ExternalJob",
+				APIVersion: "jobs.example.org/v1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: "default",
+				Name:      "obj-1",
+			},
+			Spec: jobsv1.ExternalJobSpec{
+				RunAt: "every 2 seconds",
+			},
+		}
+
+		convReq := &apix.ConversionReview{
+			TypeMeta: metav1.TypeMeta{},
+			Request: &apix.ConversionRequest{
+				// request conversion for different group
+				DesiredAPIVersion: "jobss.example.org/v2",
+				Objects: []runtime.RawExtension{
+					{
+						Object: v1Obj,
+					},
+				},
+			},
+		}
+
+		convReview := doRequest(convReq)
+		Expect(convReview.Response.Result.Message).Should(
```

Comment:
blech.  I hate directly checking error message strings.  Can't we just check codes?

Reply:
Need to explore a bit more, but in general I agree checking error messages is bad.

Reply:
I explored bit more about using codes here, couldn't find an easy way to do it at this level which is conversionRequest/Response level. At this level, we can only indicate status and failure message (which contains the reason for failure). I removed reason string checking here because that makes the tests brittle.